client/dialogs/connected-dialog-overlay.tsx

```diff
@@ -115,32 +113,30 @@ function getDialog(
 
 export function ConnectedDialogOverlay() {
   const dispatch = useAppDispatch()
-  const dialogState = useAppSelector(s => s.dialog)
+  const dialogHistory = useAppSelector(s => s.dialog.history)
+  const isDialogOpen = dialogHistory.length > 0
+  const topDialog = isDialogOpen ? dialogHistory[dialogHistory.length - 1] : undefined
   const starcraft = useAppSelector(s => s.starcraft)
 
   const focusableRef = useRef<HTMLSpanElement>(null)
   const dialogRef = useRef<HTMLElement>(null)
   const portalRef = useExternalElementRef()
 
-  const { dialogType } = dialogState
-  const { component: DialogComponent, modal } = dialogType
-    ? getDialog(dialogType, starcraft)
-    : { component: null, modal: false }
-
+  const isTopDialogModal = topDialog ? getDialog(topDialog.type, starcraft).modal : false
   const onCancel = useCallback(
-    (event?: React.MouseEvent) => {
-      if (dialogType && !modal && (!event || !isHandledDismissalEvent(event.nativeEvent))) {
+    (modal: boolean, event?: React.MouseEvent) => {
```

Comment:
This seems like a strictly worse way to structure this, I think you'd be better off adding `modal` to the dependencies array and just passing this function to the `DialogComponent` directly. There's extremely little reason to `useCallback` this function otherwise, since you're creating a new instance on every render to call it.

You'll have to move the `getDialog` call up here but I don't see anything that would prevent doing that

Reply:
I think I tried doing that initially, but not sure how that would work exactly. Unless you mean I should call `getDialog` for all currently open dialogs here to get their `modal` status, and then generate `onCancel` for each one? But then I would need to create `onCancel` in a loop or something, which I don't think can be done with hooks. Maybe I could use `useMemo`?

Reply:
If the only difference is the `modal` flag then a different option would be to generate a modal and a non-modal handler always, and pick between them at render time. Since the modal handling is always to just ignore the cancel, you could really just make that version a constant no-op function outside the component.